File: solutions/clojure/allergies/1/src/allergies.clj
(ns allergies)

(def ^:private allergens {:eggs 1, :peanuts 2, :shellfish 4, :strawberries 8,
                          :tomatoes 16, :chocolate 32, :pollen 64, :cats 128})
(def ^:private allergens-order [:eggs :peanuts :shellfish :strawberries :tomatoes :chocolate :pollen :cats])
(defn allergic-to? [score allergen] (pos? (bit-and score (allergens allergen 0))))
(defn allergies [score] (filter #(allergic-to? score %) allergens-order))

File: solutions/ruby/allergies/1/allergies.rb
# frozen_string_literal: true

KEYS = %w[eggs peanuts shellfish strawberries tomatoes chocolate pollen cats].freeze
VALUES = [1, 2, 4, 8, 16, 32, 64, 128].freeze
ALLERGENS = Hash[KEYS.zip(VALUES)]

# class Allergies
class Allergies
  def initialize(score) = @score = score
  def allergic_to?(which) = (@score & ALLERGENS[which]).positive?
  def list = KEYS.filter { allergic_to?(_1) }
end
